### PR TITLE
[OP] Lt, Softmax + dx, gelu dx

### PR DESCRIPTION
**Pull request by anadeem2**
## Description ##
Implement other missing operators used in transformer & bert like models. 
## Checklist ##

- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

cc @awslabs/raf-reviewer @zachzzc

### CODE REVIEW

File: ratex/testing/common.py [hunk 1]
```diff
@@ -374,11 +374,58 @@ def run_step(device, model_origin, args, jit_script=True):
...
+        model = ratex.jit.script(model)
+    args = [arg.to(device) for arg in args]
+    out = model(*args)
+    loss = out.sum()
+    loss.backward()
+    lm.mark_step()
+    return args
+
+
+def verify_step(model, args, jit_script=True, dx=False, tol=1e-5):
```
Comment (zachzzc) [inline ratex/testing/common.py:409]: suggest change `dx` to `with_backward`

File: ratex/testing/common.py [hunk 2]
```diff
@@ -374,11 +374,58 @@ def run_step(device, model_origin, args, jit_script=True):
...
     """Verify the results between CPU and Lazy"""
-    torch.testing.assert_close(
-        run_step("cpu", model, args), run_step("lazy", model, args, jit_script), rtol=tol, atol=tol
-    )
+    if dx:
+        args_ratex = []
+        for x in args:
```
Comment (zachzzc) [inline ratex/testing/common.py:413]: why we need to have a separate args_ratex ?

File: ratex/testing/common.py [hunk 3]
```diff
@@ -374,11 +374,58 @@ def run_step(device, model_origin, args, jit_script=True):
     return out
 
 
-def verify_step(model, args, jit_script=True, tol=1e-5):
+def run_step_dx(device, model_origin, args, jit_script=True):
```
Comment (zachzzc) [inline ratex/testing/common.py:377]: we should merge run_step and run_step_dx, with an extra argument `with_backward`.
And we should return both loss and args back.

File: ratex/testing/common.py [hunk 4]
```diff
@@ -374,11 +374,58 @@ def run_step(device, model_origin, args, jit_script=True):
...
-        run_step("cpu", model, args), run_step("lazy", model, args, jit_script), rtol=tol, atol=tol
-    )
+    if dx:
+        args_ratex = []
+        for x in args:
+            n_x = np.copy(x.data)
+            t_x_ratex = torch.tensor(n_x, device="lazy", dtype=torch.float32, requires_grad=True)
+            args_ratex.append(t_x_ratex)
+        grad_cpu = run_step_dx("cpu", model, args)
+        grad_lazy = run_step_dx("lazy", model, args_ratex, jit_script)
```
Comment (zachzzc) [inline ratex/testing/common.py:418]: Naming is confusing. This should be args_cpu and args_lazy

**Review by zachzzc — CHANGES_REQUESTED** [review]

File: tests/python/op/test_nn.py [hunk 5]
```diff
@@ -86,6 +88,21 @@ def forward(self, x):
     verify_step(Model(), [x], jit_script=False)
 
 
+def test_gelu_backward():
```
[the remark this answers is not in the document]
Comment (anadeem2) [reply tests/python/op/test_nn.py:91]: gelu kernel is not implemented for float16/half.

File: tests/python/op/test_nn.py [hunk 6]
```diff
@@ -101,5 +118,34 @@ def forward(self, x_input):
...
+
+        def forward(self, x_input):
+            return self.softmax(x_input)
+
+    x = torch.randn(3, 3)
+
+    verify_step(Model(), [x], jit_script=False)
+
+
+def test_softmax_backward():
```
[the remark this answers is not in the document]
Comment (anadeem2) [reply tests/python/op/test_nn.py:135]: softmax kernel is not implemented for float16/half.

File: ratex/testing/common.py (hunk 2 above)
Comment (anadeem2) [reply ratex/testing/common.py:413]: As you can see, we have 1 reference to args. So in this case, we have 1 tensor that first gets its grad from cpu, but then we then re-forward/auto diff w/ lazy so it gets another grad. This means CPU grad becomes None, so we need to copy. Also, for some reason, lazy requires_grad=True does not work w/ regular torch.randn, so we need to use numpy then create tensor.

**Comment by zachzzc** [comment]
@anadeem2  Can you resolve the conflicts?

File: ratex/csrc/aten_raf_type.cpp [hunk 7]
```diff
@@ -440,6 +440,7 @@ at::Tensor LazyNativeFunctions::where(const at::Tensor& condition, const at::Ten
 at::Tensor LazyNativeFunctions::_softmax(const at::Tensor& self, int64_t dim,
                                          bool /* half_to_float */) {
   LTC_FN_COUNTER("raf::");
+  LTC_CHECK_EQ(dim, self.dim() - 1);
```
Comment (zachzzc) [inline ratex/csrc/aten_raf_type.cpp:443]: what does self.dim() - 1 mean? why not just compare with -1 ?

**Review by zachzzc — APPROVED** [review]
LGTM

**Review by zachzzc — APPROVED** [review]
LGTM

**Comment by zachzzc** [comment]
Thanks @anadeem2
